lib/rubocop/cop/rspec/subject_stub.rb

```diff
@@ -74,71 +76,54 @@ class SubjectStub < Cop
             add_offense(stub)
           end
         end
 
         private
 
-        # Find subjects within tree and then find (send) nodes for that subject
-        #
-        # @param node [RuboCop::Node] example group
-        #
-        # @yield [RuboCop::Node] message expectations for subject
-        def find_subject_stub(node, &block)
-          find_subject(node) do |subject_name, context|
-            find_subject_expectation(context, subject_name, &block)
-          end
+        def processed_example_groups
+          @processed_example_groups ||= Set[]
         end
 
-        # Find a subject message expectation
-        #
-        # @param node [RuboCop::Node]
-        # @param subject_name [Symbol] name of subject
-        #
-        # @yield [RuboCop::Node] message expectation
-        def find_subject_expectation(node, subject_name, &block)
-          # Do not search node if it is an example group with its own subject.
-          return if example_group?(node) && redefines_subject?(node)
-
-          # Yield the current node if it is a message expectation.
-          yield(node) if message_expectation?(node, subject_name)
-
-          # Recurse through node's children looking for a message expectation.
-          node.each_child_node do |child|
-            find_subject_expectation(child, subject_name, &block)
+        def find_all_named_subjects(node)
+          node.each_descendant(:block).each_with_object({}) do |child, h|
+            name = subject(child)
+            h[child.parent.parent] = name if name
           end
         end
 
-        # Check if node's children contain a subject definition
-        #
-        # @param node [RuboCop::Node]
-        #
-        # @return [Boolean]
-        def redefines_subject?(node)
-          node.each_child_node.any? do |child|
-            subject(child) || redefines_subject?(child)
+        def find_subject_expectations(node, subject_name = nil, &block)
+          # if it's a new example group - check whether new named subject is
+          # defined there
+          if example_group?(node)
+            subject_name = @named_subjects[node] || subject_name
           end
-        end
 
-        # Find a subject definition
-        #
-        # @param node [RuboCop::Node]
-        # @param parent [RuboCop::Node,nil]
-        #
-        # @yieldparam subject_name [Symbol] name of subject being defined
-        # @yieldparam parent [RuboCop::Node] parent of subject definition
-        def find_subject(node, parent: nil, &block)
-          # An implicit subject is defined by RSpec when no subject is declared
-          subject_name = subject(node) || :subject
+          # check default :subject and then named one (if it's present)
+          expectation_detected = message_expectation?(node, :subject) || \
+            (subject_name && message_expectation?(node, subject_name))
 
-          yield(subject_name, parent) if parent
+          return yield(node) if expectation_detected
```

Comment:
There could be several subjects defined if I'm not mistaken:
```ruby
subject { ... } # unnamed
subject(:cop) { ... }
subject(:code) { ... }
subject(:performance) { ... }
```
in the same context. Are we handling this situation, do we detect attempts to stub all of them?

Reply:
Good point. Didn't think about this case.

Reply:
Fixed